websocket/connection.go

```diff
@@ -154,50 +161,53 @@ func (c *ManagedConnection) connect() (err error) {
 	return err
 }
 
-// keepalive keeps the connection open and reads control messages.
-// All messages are discarded.
-func (c *ManagedConnection) keepalive() (err error) {
+// keepalive keeps the connection open.
+func (c *ManagedConnection) keepalive() error {
+	for {
+		if err := c.read(); err != nil {
+			return err
+		}
+	}
+}
+
+// read reads the next message from the connection.
+// If a messageChan is supplied and the current message type is not
+// a control message, the message is sent to that channel.
+func (c *ManagedConnection) read() error {
+	c.connectionLock.Lock()
+	defer c.connectionLock.Unlock()
+
+	if c.connection == nil {
+		return ErrConnectionNotEstablished
+	}
+
 	c.readerLock.Lock()
 	defer c.readerLock.Unlock()
 
-	for {
-		func() {
-			c.connectionLock.RLock()
-			defer c.connectionLock.RUnlock()
-
-			if conn := c.connection; conn != nil {
-				var reader io.Reader
-				var messageType int
-				messageType, reader, err = conn.NextReader()
-				if err != nil {
-					conn.Close()
-				}
-
-				// Send the message to the channel if its an application level message
-				// and if that channel is set.
-				if c.messageChan != nil && (messageType == websocket.TextMessage || messageType == websocket.BinaryMessage) {
-					if message, _ := ioutil.ReadAll(reader); message != nil {
-						c.messageChan <- message
-					}
-				}
-			} else {
-				err = ErrConnectionNotEstablished
-			}
-		}()
+	messageType, reader, err := c.connection.NextReader()
+	if err != nil {
```

Comment:
this block is repeated at least twice. Should probably be a function.

Reply:
Hm, I see what you mean. However the only part we can put into a function is

```golang
c.connection.Close()
c.connection = nil
```

And they theoretically need to hold a lock (which we'd need the caller to do). I'd prefer to keep the access to `c.connection` local to the functions they are in today for it makes clear that the lock is held by that function and thus this access is safe. WDYT?

Reply:
SG.